File: js/src/browser/browser-cookie-crypto.js
import {
  createDecipheriv,
  createHash,
  pbkdf2Sync,
  timingSafeEqual,
} from 'node:crypto';
import { TextDecoder } from 'node:util';

const CHROMIUM_CBC_IV = Buffer.alloc(16, 0x20);
const DOMAIN_HASH_BYTES = 32;

export function deriveChromiumCookieKey(password, platform) {
  if (platform !== 'darwin' && platform !== 'linux') {
    throw new Error(`CBC cookie keys are not used on ${platform}`);
  }
  return pbkdf2Sync(
    password,
    'saltysalt',
    platform === 'darwin' ? 1003 : 1,
    16,
    'sha1'
  );
}

function removeDomainHash(plaintext, host, databaseVersion) {
  if (databaseVersion < 24) {
    return plaintext;
  }
  if (plaintext.length < DOMAIN_HASH_BYTES) {
    throw new Error('decrypted cookie is missing its domain hash');
  }
  const expectedHash = createHash('sha256').update(host).digest();
  const actualHash = plaintext.subarray(0, DOMAIN_HASH_BYTES);
  if (!timingSafeEqual(actualHash, expectedHash)) {
    throw new Error('decrypted cookie domain hash does not match its host');
  }
  return plaintext.subarray(DOMAIN_HASH_BYTES);
}

function decodeCookieValue(plaintext) {
  return new TextDecoder('utf-8', { fatal: true }).decode(plaintext);
}

/** Validate and decode plaintext returned by a Chromium platform decryptor. */
export function decodeChromiumCookiePlaintext({
  plaintext,
  host,
  databaseVersion = 0,
}) {
  return decodeCookieValue(
    removeDomainHash(Buffer.from(plaintext), host, databaseVersion)
  );
}

function decryptCbcCookie(encryptedValue, key) {
  const decipher = createDecipheriv('aes-128-cbc', key, CHROMIUM_CBC_IV);
  return Buffer.concat([
    decipher.update(encryptedValue.subarray(3)),
    decipher.final(),
  ]);
}

function decryptGcmCookie(encryptedValue, key) {
  const payload = encryptedValue.subarray(3);
  if (payload.length < 12 + 16) {
    throw new Error('encrypted AES-GCM cookie is truncated');
  }
  const nonce = payload.subarray(0, 12);
  const authenticationTag = payload.subarray(-16);
  const ciphertext = payload.subarray(12, -16);
  const decipher = createDecipheriv('aes-256-gcm', key, nonce);
  decipher.setAuthTag(authenticationTag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

/** Decrypt a Chromium cookie encrypted_value using an already recovered key. */
export function decryptChromiumCookie({
  encryptedValue,
  host,
  databaseVersion = 0,
  platform = process.platform,
  key,
}) {
  const encrypted = Buffer.from(encryptedValue);
  const prefix = encrypted.subarray(0, 3).toString('ascii');
  if (prefix === 'v20') {
    throw new Error(
      'Windows app-bound v20 cookies cannot be decrypted outside the browser; use a browser-supported export or a previously saved storage state'
    );
  }
  if (prefix !== 'v10' && prefix !== 'v11') {
    throw new Error(
      'cookie does not use a supported Chromium encryption prefix'
    );
  }
  if (!Buffer.isBuffer(key)) {
    throw new TypeError('a recovered cookie encryption key is required');
  }
  const plaintext =
    platform === 'win32'
      ? decryptGcmCookie(encrypted, key)
      : decryptCbcCookie(encrypted, key);
  return decodeChromiumCookiePlaintext({
    plaintext,
    host,
    databaseVersion,
  });
}

export function chromiumSameSite(value) {
  if (Number(value) === 2) {
    return 'Strict';
  }
  if (Number(value) === 1) {
    return 'Lax';
  }
  return 'None';
}

export function firefoxSameSite(value) {
  if (Number(value) === 2) {
    return 'Strict';
  }
  if (Number(value) === 1) {
    return 'Lax';
  }
  return 'None';
}

File: rust/src/browser/browser_cookie_crypto.rs
//! Chromium cookie key derivation and decryption primitives.

use aes::Aes128;
use aes_gcm::aead::{Aead, KeyInit};
use aes_gcm::{Aes256Gcm, Nonce};
use anyhow::{anyhow, Context, Result};
use cbc::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
use pbkdf2::pbkdf2_hmac;
use sha1::Sha1;
use sha2::{Digest, Sha256};

type Aes128CbcDecryptor = cbc::Decryptor<Aes128>;

pub(crate) fn derive_chromium_cookie_key(password: &str, platform: &str) -> Result<Vec<u8>> {
    let iterations = match platform {
        "darwin" => 1003,
        "linux" => 1,
        _ => return Err(anyhow!("CBC cookie keys are not used on {platform}")),
    };
    let mut key = vec![0_u8; 16];
    pbkdf2_hmac::<Sha1>(password.as_bytes(), b"saltysalt", iterations, &mut key);
    Ok(key)
}

fn remove_domain_hash<'a>(plaintext: &'a [u8], host: &str, version: i64) -> Result<&'a [u8]> {
    if version < 24 {
        return Ok(plaintext);
    }
    if plaintext.len() < 32 {
        return Err(anyhow!("decrypted cookie is missing its domain hash"));
    }
    let expected = Sha256::digest(host.as_bytes());
    if plaintext[..32] != expected[..] {
        return Err(anyhow!(
            "decrypted cookie domain hash does not match its host"
        ));
    }
    Ok(&plaintext[32..])
}

fn decrypt_cbc(encrypted: &[u8], key: &[u8]) -> Result<Vec<u8>> {
    Aes128CbcDecryptor::new_from_slices(key, &[0x20; 16])
        .context("invalid Chromium AES-CBC key")?
        .decrypt_padded_vec_mut::<Pkcs7>(&encrypted[3..])
        .map_err(|_| anyhow!("Chromium AES-CBC cookie padding is invalid"))
}

fn decrypt_gcm(encrypted: &[u8], key: &[u8]) -> Result<Vec<u8>> {
    let payload = &encrypted[3..];
    if payload.len() < 28 {
        return Err(anyhow!("encrypted AES-GCM cookie is truncated"));
    }
    let cipher = Aes256Gcm::new_from_slice(key).context("invalid Chromium AES-GCM key")?;
    cipher
        .decrypt(Nonce::from_slice(&payload[..12]), &payload[12..])
        .map_err(|_| anyhow!("Chromium AES-GCM cookie authentication failed"))
}

pub(crate) fn decrypt_chromium_cookie(
    encrypted: &[u8],
    host: &str,
    database_version: i64,
    platform: &str,
    key: &[u8],
) -> Result<String> {
    if encrypted.len() < 3 {
        return Err(anyhow!("encrypted Chromium cookie is truncated"));
    }
    match &encrypted[..3] {
        b"v20" => {
            return Err(anyhow!(
                "Windows app-bound v20 cookies cannot be decrypted outside the browser; use a browser-supported export or a previously saved storage state"
            ))
        }
        b"v10" | b"v11" => {}
        _ => return Err(anyhow!("cookie has an unsupported Chromium encryption prefix")),
    }
    let plaintext = if platform == "win32" {
        decrypt_gcm(encrypted, key)?
    } else {
        decrypt_cbc(encrypted, key)?
    };
    String::from_utf8(remove_domain_hash(&plaintext, host, database_version)?.to_vec())
        .context("decrypted cookie is not valid UTF-8")
}

pub(crate) fn chromium_same_site(value: i64) -> &'static str {
    match value {
        2 => "Strict",
        1 => "Lax",
        _ => "None",
    }
}

pub(crate) fn firefox_same_site(value: i64) -> &'static str {
    chromium_same_site(value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use aes_gcm::aead::Aead;
    use sha2::Sha256;

    #[test]
    fn decrypts_windows_gcm_and_rejects_app_bound_cookies() {
        let key = [7_u8; 32];
        let host = ".example.test";
        let mut plaintext = Sha256::digest(host.as_bytes()).to_vec();
        plaintext.extend_from_slice(b"windows-session");
        let nonce = [3_u8; 12];
        let ciphertext = Aes256Gcm::new_from_slice(&key)
            .unwrap()
            .encrypt(Nonce::from_slice(&nonce), plaintext.as_slice())
            .unwrap();
        let encrypted = [b"v10".as_slice(), &nonce, ciphertext.as_slice()].concat();
        assert_eq!(
            decrypt_chromium_cookie(&encrypted, host, 24, "win32", &key).unwrap(),
            "windows-session"
        );
        assert!(
            decrypt_chromium_cookie(b"v20app-bound", host, 24, "win32", &key)
                .unwrap_err()
                .to_string()
                .contains("app-bound")
        );
    }

    #[test]
    fn legacy_dpapi_plaintext_validates_version_24_host_hash() {
        let host = ".legacy.example";
        let mut plaintext = Sha256::digest(host.as_bytes()).to_vec();
        plaintext.extend_from_slice(b"legacy-session");
        assert_eq!(
            decode_chromium_plaintext(&plaintext, host, 24).unwrap(),
            "legacy-session"
        );
        assert!(decode_chromium_plaintext(&plaintext, ".wrong.example", 24)
            .unwrap_err()
            .to_string()
            .contains("domain hash does not match"));
    }
}
